### PR TITLE
Special case Rational * 1 and 1 * Rational

### PR DESCRIPTION
**Pull request by isuruf**
Avoids a call to igcd which takes a lot of time if the numerator and denominator of the Rational are large.

cc @rwst

### CODE REVIEW

**Comment by rwst** [comment]
+1

**Comment by isuruf** [comment]
Tests pass. Ready for review now

**Comment by jksuom** [comment]
Looks good to me apart from a couple of remarks:

The tests for ``S.One`` could be in the beginning, before ``if isinstance(other, Rational):``.

``self is S.One`` will probably never occur in ``Rational.__mul__``. Instead, it could be in ``Integer.__mul__``.

**Comment by isuruf** [comment]
I did try the tests before the `if isinstance(other, Rational):`, but I cannot return other if self is `S.One` because `1 * 0.0` should return `0`

`self is S.One` occurs in `Integer.__mul__` and if `other` is not a integer type then `Rational.__mul__` is called

**Comment by jksuom** [comment]
> 1 * 0.0 should return 0

I think this is odd. Is there a compelling reason for change of type?

**Comment by isuruf** [comment]
Actually no. When we implemented floats in SymEngine we made sure to return 0.0 which I think is the best approach. 

In SymPy however many tests depend on the fact that 1*0.0 = 0.

**Comment by jksuom** [comment]
Thanks, I didn't know that. (I've seen that floats behave unexpectedly in some other connections.)

**Comment by smichr** [comment]
> Is there a compelling reason for change of type?

I'm not sure that we can track the precision of Float zero. So there's no point in keeping a floating point 0 instead of a plain 0 (I think).

**Comment by smichr** [comment]
>  I cannot return other if self is  S.One  because  1 * 0.0  should return  0 

you could return `other or S.Zero` which, if other is 0.0 will return the non-Float zero.

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -1389,7 +1389,12 @@ def __sub__(self, other):
     @_sympifyit('other', NotImplemented)
     def __mul__(self, other):
         if isinstance(other, Rational):
```
Comment (smichr) [inline sympy/core/numbers.py]:
```
if self is S.One:
    other, self = self, other
if other is S.One:
    return self or S.Zero
if isinstance(other, Rational):
    return Rational(self.p*other.p, self.q*other.q) 
...
```

**Comment by smichr** [comment]
Sorry, I didn't know about that restriction on `self`. So how about

```python
    if self is S.One:
        return other or S.Zero
    if other is S.One:
        return self or S.Zero
    ...
```

**Comment by jksuom** [comment]
+1

**Comment by smichr** [comment]
I have no idea why the ode test is failing. I've added a few changes and am testing it now.

**Comment by smichr** [comment]
minor mods [here](https://github.com/isuruf/sympy/pull/1)

**Comment by smichr** [comment]
I saved the log if anyone wants to see it. I then restarted the job.

**Comment by isuruf** [comment]
Closed via #11019

**Pull request closed without merging**